From 6d79ef27abed0d8eecae05c8c29708a9104b4728 Mon Sep 17 00:00:00 2001
From: Matthew Gharrity <gharrma@google.com>
Date: Fri, 28 Jul 2023 23:20:25 +0000
Subject: [PATCH] [do not merge] Bump Compose Kotlinc version to
 1.9.20-dev-5606

In order to update to IntelliJ 2023.2 in Android Studio, we need to
bundle a Compose compiler that is compatible with the Kotlin compiler
shipped in IntelliJ (version 1.9.20-dev-5606). Otherwise both
Compose editing support and LiveEdit break due to binary
incompatibilities.

This change is here to demonstrate the changes needed to create
a compatible Compose build.

Change-Id: I985707239a0b35feb048da1436a1f3f04f72f39d
---
 buildSrc/repos.gradle                            |  3 +++
 compose/compiler/compiler-hosted/build.gradle    |  2 +-
 .../kotlin/k2/ComposableFunctionChecker.kt       |  4 ++--
 .../lower/ComposableFunctionBodyTransformer.kt   |  2 +-
 .../plugins/kotlin/lower/IrSourcePrinter.kt      |  2 +-
 .../kotlin/lower/decoys/DecoyTransformBase.kt    |  3 ++-
 gradle/verification-metadata.xml                 | 16 ++++++++++++++++
 7 files changed, 26 insertions(+), 6 deletions(-)

diff --git a/buildSrc/repos.gradle b/buildSrc/repos.gradle
index 9b5077d0f8b..bd178e334cf 100644
--- a/buildSrc/repos.gradle
+++ b/buildSrc/repos.gradle
@@ -57,6 +57,9 @@ def addMavenRepositories(RepositoryHandler handler) {
             artifact()
         }
     }
+    handler.maven {
+        url("https://maven.pkg.jetbrains.space/kotlin/p/kotlin/kotlin-ide-plugin-dependencies")
+    }
     if (System.getenv("ALLOW_PUBLIC_REPOS") != null || System.getProperty("ALLOW_PUBLIC_REPOS") != null) {
         handler.mavenCentral()
         handler.google()
diff --git a/compose/compiler/compiler-hosted/build.gradle b/compose/compiler/compiler-hosted/build.gradle
index e4ce28e71d4..c2a1ca1594b 100644
--- a/compose/compiler/compiler-hosted/build.gradle
+++ b/compose/compiler/compiler-hosted/build.gradle
@@ -25,7 +25,7 @@ plugins {
 
 dependencies {
     compileOnly(libs.kotlinStdlib)
-    compileOnly(libs.kotlinCompiler)
+    compileOnly("org.jetbrains.kotlin:kotlin-compiler:1.9.20-dev-5606")
 
     testImplementation(libs.kotlinStdlib)
     testImplementation(libs.junit)
diff --git a/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/k2/ComposableFunctionChecker.kt b/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/k2/ComposableFunctionChecker.kt
index 677daeceb6b..4e2e8074814 100644
--- a/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/k2/ComposableFunctionChecker.kt
+++ b/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/k2/ComposableFunctionChecker.kt
@@ -22,7 +22,7 @@ import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
 import org.jetbrains.kotlin.fir.analysis.checkers.declaration.FirFunctionChecker
 import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
 import org.jetbrains.kotlin.fir.declarations.FirFunction
-import org.jetbrains.kotlin.fir.declarations.getSingleCompatibleExpectForActualOrNull
+import org.jetbrains.kotlin.fir.declarations.getSingleExpectForActualOrNull
 import org.jetbrains.kotlin.fir.declarations.utils.isAbstract
 import org.jetbrains.kotlin.fir.declarations.utils.isOperator
 import org.jetbrains.kotlin.fir.declarations.utils.isSuspend
@@ -52,7 +52,7 @@ object ComposableFunctionChecker : FirFunctionChecker() {
         }
 
         // Check that `actual` composable declarations have composable expects
-        declaration.symbol.getSingleCompatibleExpectForActualOrNull()?.let { expectDeclaration ->
+        declaration.symbol.getSingleExpectForActualOrNull()?.let { expectDeclaration ->
             if (expectDeclaration.hasComposableAnnotation(context.session) != isComposable) {
                 reporter.reportOn(
                     declaration.source,
diff --git a/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/ComposableFunctionBodyTransformer.kt b/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/ComposableFunctionBodyTransformer.kt
index e3348596ed0..582b66ade45 100644
--- a/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/ComposableFunctionBodyTransformer.kt
+++ b/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/ComposableFunctionBodyTransformer.kt
@@ -3942,7 +3942,7 @@ class ComposableFunctionBodyTransformer(
                 var parent = function.parent
                 while (true) {
                     when (parent) {
-                        is IrPackageFragment -> return parent.fqName.asString()
+                        is IrPackageFragment -> return parent.packageFqName.asString()
                         is IrDeclaration -> parent = parent.parent
                         else -> break
                     }
diff --git a/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/IrSourcePrinter.kt b/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/IrSourcePrinter.kt
index 04c7f2c21ef..d72d3175db6 100644
--- a/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/IrSourcePrinter.kt
+++ b/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/IrSourcePrinter.kt
@@ -1520,7 +1520,7 @@ class IrSourcePrinterVisitor(
             if (parent is IrDeclaration) {
                 parent.renderDeclarationFqn(sb)
             } else if (parent is IrPackageFragment) {
-                sb.append(parent.fqName.toString())
+                sb.append(parent.packageFqName.toString())
             }
         } catch (e: UninitializedPropertyAccessException) {
             sb.append("<uninitialized parent>")
diff --git a/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/decoys/DecoyTransformBase.kt b/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/decoys/DecoyTransformBase.kt
index 632e6b00eeb..bb62512ef39 100644
--- a/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/decoys/DecoyTransformBase.kt
+++ b/compose/compiler/compiler-hosted/src/main/java/androidx/compose/compiler/plugins/kotlin/lower/decoys/DecoyTransformBase.kt
@@ -117,7 +117,8 @@ internal interface DecoyTransformBase {
             packageFqName = signature[0],
             declarationFqName = signature[1],
             id = signature[2].toLongOrNull(),
-            mask = signature[3].toLong()
+            mask = signature[3].toLong(),
+            null, // TODO: "set 'description' to the mangled name from which 'id' was computed, or to null if it's not applicable"
         )
 
         val linker = (context as IrPluginContextImpl).linker
diff --git a/gradle/verification-metadata.xml b/gradle/verification-metadata.xml
index 2c9bedec632..0da686c5f86 100644
--- a/gradle/verification-metadata.xml
+++ b/gradle/verification-metadata.xml
@@ -463,6 +463,22 @@
       </trusted-keys>
    </configuration>
    <components>
+      <component group="org.jetbrains.kotlin" name="kotlin-compiler" version="1.9.20-dev-5606">
+         <artifact name="kotlin-compiler-1.9.20-dev-5606.pom">
+            <sha256 value="35155404483a8a5983ddec6a74f98a3694b404b63402814771e887f93ad9b71d" origin="Hand-built using sha256sum kotlin-compiler-1.9.20-dev-5606.pom" reason="Artifact is not signed"/>
+         </artifact>
+         <artifact name="kotlin-compiler-1.9.20-dev-5606.jar">
+            <sha256 value="7ad09ce74fc738af2d286338cee47ac624272fb3e2971145f1d5cb234e9871e4" origin="Hand-built using sha256sum kotlin-compiler-1.9.20-dev-5606.jar" reason="Artifact is not signed"/>
+         </artifact>
+      </component>
+      <component group="org.jetbrains.kotlin" name="kotlin-script-runtime" version="1.9.20-dev-5606">
+         <artifact name="kotlin-script-runtime-1.9.20-dev-5606.pom">
+            <sha256 value="0f42b1e9757d2ce097156abb6e0c8608636a626c203f116394cbffe8a98b141b" origin="Hand-built using sha256sum kotlin-script-runtime-1.9.20-dev-5606.pom" reason="Artifact is not signed"/>
+         </artifact>
+         <artifact name="kotlin-script-runtime-1.9.20-dev-5606.jar">
+            <sha256 value="98a629b8f67ae72e46e5356198ed1cd7c46ef51144e0f085fb911aa50e9338fe" origin="Hand-built using sha256sum kotlin-script-runtime-1.9.20-dev-5606.jar" reason="Artifact is not signed"/>
+         </artifact>
+      </component>
       <component group="" name="kotlin-native-prebuilt-linux-x86_64" version="1.9.0">
          <artifact name="kotlin-native-prebuilt-linux-x86_64-1.9.0.tar.gz">
             <sha256 value="31737a9739fc37208e1f532b7472c3fbbf0d753f3621c9dfc1d72a69d5bc35c0" origin="Hand-built using sha256sum kotlin-native-prebuilt-linux-x86_64-1.9.0.tar.gz" reason="Artifact is not signed"/>
-- 
2.41.0.640.ga95def55d0-goog

